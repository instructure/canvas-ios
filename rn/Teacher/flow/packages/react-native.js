//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

declare var console: typeof console & { disableYellowBox: boolean }

export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}

export type KeyboardEventData = {
  endCoordinates: {
    width: number,
    height: number,
    screenX: number,
    screenY: number,
  },
}
