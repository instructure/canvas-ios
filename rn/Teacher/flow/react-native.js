//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow
/* eslint-disable */

// define basic style prop types
// the internal ones are the best we have
import {
  type ____ViewStyleProp_Internal,
  type ____TextStyleProp_Internal,
  type ____ImageStyleProp_Internal,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes'

declare type ViewStyleProp = ____ViewStyleProp_Internal
declare type TextStyleProp = ____TextStyleProp_Internal
declare type ImageStyleProp = ____ImageStyleProp_Internal
