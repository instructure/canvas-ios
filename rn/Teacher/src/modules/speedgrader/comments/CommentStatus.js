//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import React, { Component } from 'react'
import {
  View,
  Dimensions,
  Animated,
  Easing,
} from 'react-native'
import DrawerState from '../utils/drawer-state'
import { createStyleSheet } from '../../../common/stylesheet'

const DEVICE_WIDTH = Dimensions.get('window').width

type CommentStatusState = {
  width: number,
}

type CommentStatusProps = {
  isDone: boolean,
  animationComplete: Function,
  drawerState: DrawerState,
  userID: string,
}

export default class CommentStatus extends Component<CommentStatusProps, CommentStatusState> {
  state: CommentStatusState = { width: DEVICE_WIDTH }
  progress: Animated.Value = this.props.drawerState.commentProgress[this.props.userID] || new Animated.Value(0)

  constructor (props: CommentStatusProps) {
    super(props)
    this.props.drawerState.registerCommentProgress(this.props.userID, this.progress)
  }

  componentDidMount () {
    let duration = 60000 * (1 - this.progress._value)
    const animation = Animated.timing(
      this.progress,
      {
        toValue: 0.8,
        duration,
        easing: Easing.linear,
        useNativeDriver: false,
      },
    )
    if (animation) {
      animation.start()
    }
  }

  componentWillUnmount () {
    this.props.drawerState.unregisterCommentProgress(this.props.userID)
  }

  UNSAFE_componentWillReceiveProps (nextProps: CommentStatusProps) {
    if (nextProps.isDone) {
      this.progress.stopAnimation()
      Animated.timing(
        this.progress,
        {
          toValue: 1,
          duration: 300,
          useNativeDriver: false,
        }
      ).start(nextProps.animationComplete)
    }
  }

  onLayout = (e: any) => {
    this.setState({
      width: e.nativeEvent.layout.width,
    })
  }

  render () {
    return (
      <View style={styles.statusBar} onLayout={this.onLayout}>
        <Animated.View
          style={[
            styles.progress,
            {
              width: this.progress.interpolate({
                inputRange: [0, 1],
                outputRange: [0, this.state.width],
              }),
            },
          ]}
        />
      </View>
    )
  }
}

const styles = createStyleSheet(colors => ({
  statusBar: {
    position: 'absolute',
    bottom: 50,
    left: 0,
    right: 0,
    height: 1,
    flexDirection: 'row',
  },
  progress: {
    backgroundColor: colors.backgroundInfo,
  },
}))
