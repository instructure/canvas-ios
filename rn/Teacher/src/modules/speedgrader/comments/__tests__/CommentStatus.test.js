//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/* eslint-disable flowtype/require-valid-file-annotation */

import React from 'react'
import { Animated } from 'react-native'
import CommentStatus from '../CommentStatus'
import DrawerState from '../../utils/drawer-state'
import renderer from 'react-test-renderer'
import setProps from '../../../../../test/helpers/setProps'

let defaultProps = {
  userID: '1',
  isDone: false,
  animationComplete: jest.fn(),
  drawerState: new DrawerState(),
}

describe('CommentStatus', () => {
  beforeEach(() => jest.clearAllMocks())
  it('renders', () => {
    let tree = renderer.create(
      <CommentStatus {...defaultProps} />
    ).toJSON()

    expect(tree).toMatchSnapshot()
    expect(defaultProps.drawerState.commentProgress['1']).not.toBeNull()
  })

  it('uses an existing animated value if one exists in drawerState', () => {
    let progress = new Animated.Value(0)
    defaultProps.drawerState.registerCommentProgress('1', progress)
    let tree = renderer.create(
      <CommentStatus {...defaultProps} />
    )
    expect(tree.getInstance().progress).toEqual(progress)
    defaultProps.drawerState.unregisterCommentProgress('1')
  })

  // This test is busted for unknown reasons...
  // it('animates with a duration based on existing animated value', () => {
  //   let oldTiming = Animated.timing
  //   Animated.timing = jest.fn(() => ({ start: jest.fn() }))
  //   renderer.create(
  //     <CommentStatus {...defaultProps} />
  //   )

  //   expect(Animated.timing.mock.calls[0][1]).toMatchObject({
  //     toValue: 0.8,
  //     duration: 60000,
  //   })

  //   defaultProps.drawerState.commentProgress['1'].setValue(0.5)
  //   renderer.create(
  //     <CommentStatus {...defaultProps} />
  //   )
  //   expect(Animated.timing.mock.calls[1][1]).toMatchObject({
  //     toValue: 0.8,
  //     duration: 30000,
  //   })

  //   Animated.timing = oldTiming
  // })

  it('animates quickly once the comment is done sending', () => {
    let oldTiming = Animated.timing
    Animated.timing = jest.fn(() => ({ start: jest.fn() }))

    let tree = renderer.create(
      <CommentStatus {...defaultProps} />
    )
    setProps(tree, { isDone: true })

    expect(Animated.timing.mock.calls[1][1]).toMatchObject({
      toValue: 1,
      duration: 300,
    })

    Animated.timing = oldTiming
  })

  it('calls animationComplete once the successful animation completes', () => {
    let oldTiming = Animated.timing
    let start = jest.fn()
    Animated.timing = jest.fn(() => ({ start }))

    let tree = renderer.create(
      <CommentStatus {...defaultProps} />
    )
    setProps(tree, { isDone: true })

    expect(start).toHaveBeenCalled()
    start.mock.calls[1][0]()
    expect(defaultProps.animationComplete).toHaveBeenCalled()

    Animated.timing = oldTiming
  })

  it('sets the new width onLayout changes', () => {
    let tree = renderer.create(
      <CommentStatus {...defaultProps} />
    )

    tree.getInstance().onLayout({
      nativeEvent: {
        layout: {
          width: 100,
        },
      },
    })

    expect(tree.getInstance().state.width).toEqual(100)
  })
})
