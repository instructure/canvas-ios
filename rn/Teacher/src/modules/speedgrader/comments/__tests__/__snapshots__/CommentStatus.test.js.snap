// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentStatus renders 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "bottom": 50,
      "flexDirection": "row",
      "height": 1,
      "left": 0,
      "position": "absolute",
      "right": 0,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#008EE2",
        "width": 0,
      }
    }
  />
</View>
`;
